File: hej/apps/hej-web/lib/api/task-history.ts
/**
 * Task history API: audit activity stream and derived workflow lineage.
 */
import { apiClient, type ApiResult } from "./client"

export type HistoryCategory =
  | "setup"
  | "policy"
  | "data"
  | "annotation"
  | "review"
  | "dispute"
  | "system"

export type HistoryActor = {
  kind: "user" | "system"
  id: number | null
  display_name: string
}

export type HistoryResource = {
  type: string
  id: string
  label?: string | null
}

export type HistoryChange = {
  field: string
  before: unknown
  after: unknown
}

export type TaskHistoryEvent = {
  id: number
  occurred_at: string
  category: HistoryCategory
  operation: string
  summary: string
  detail: string | null
  actor: HistoryActor
  resource: HistoryResource | null
  changes: HistoryChange[]
}

export type TaskAuditLogsResponse = {
  task_id: string
  project_id: string | null
  total_count: number
  logs: TaskHistoryEvent[]
}

export type TaskItemCounts = {
  total: number
  pending: number
  annotated: number
  reviewed: number
  disputed: number
  canonicalized: number
}

export type WorkflowStepState =
  | "pending"
  | "in_progress"
  | "complete"
  | "blocked"
  | "later"

export type WorkflowStep = {
  key: string
  title: string
  description: string
  state: WorkflowStepState
  tone: "default" | "accent" | "dark"
  evidence: string | null
  filter_categories?: HistoryCategory[]
  workspace_tab?: string | null
}

export type TaskWorkflowLineageResponse = {
  task_id: string
  task_status: string
  annotation_mode: string
  generated_at: string
  item_counts: TaskItemCounts
  open_escalations: number
  steps: WorkflowStep[]
  current_step_key: string | null
  headline?: string
  next_action?: string
}

export type ListTaskAuditLogsParams = {
  limit?: number
  offset?: number
  category?: HistoryCategory
}

export function getTaskAuditLogs(
  taskId: string,
  params: ListTaskAuditLogsParams = {},
): Promise<ApiResult<TaskAuditLogsResponse>> {
  const search = new URLSearchParams()
  if (params.limit != null) search.set("limit", String(params.limit))
  if (params.offset != null) search.set("offset", String(params.offset))
  if (params.category) search.set("category", params.category)
  const query = search.toString()
  return apiClient.get<TaskAuditLogsResponse>(
    `/tasks/${taskId}/audit-logs${query ? `?${query}` : ""}`,
  )
}

export function getTaskWorkflowLineage(
  taskId: string,
): Promise<ApiResult<TaskWorkflowLineageResponse>> {
  return apiClient.get<TaskWorkflowLineageResponse>(
    `/tasks/${taskId}/workflow-lineage`,
  )
}

File: hej/apps/hej-api/app/schemas/task_history.py
"""Task-scoped history: enriched audit logs and derived workflow lineage."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field

HistoryCategory = Literal[
    "setup", "policy", "data", "annotation", "review", "dispute", "system"
]

WorkflowStepState = Literal[
    "pending", "in_progress", "complete", "blocked", "later"
]

WorkflowStepTone = Literal["default", "accent", "dark"]


class HistoryActorRead(BaseModel):
    kind: Literal["user", "system"]
    id: int | None = None
    display_name: str


class HistoryResourceRead(BaseModel):
    type: str
    id: str
    label: str | None = None


class HistoryChangeRead(BaseModel):
    field: str
    before: Any = None
    after: Any = None


class TaskHistoryEventRead(BaseModel):
    id: int
    occurred_at: datetime
    category: HistoryCategory
    operation: str
    summary: str
    detail: str | None = None
    actor: HistoryActorRead
    resource: HistoryResourceRead | None = None
    changes: list[HistoryChangeRead] = Field(default_factory=list)


class TaskAuditLogsResponse(BaseModel):
    task_id: str
    project_id: str | None = None
    total_count: int
    logs: list[TaskHistoryEventRead]


class TaskItemCountsRead(BaseModel):
    total: int = 0
    pending: int = 0
    annotated: int = 0
    reviewed: int = 0
    disputed: int = 0
    canonicalized: int = 0


class WorkflowStepRead(BaseModel):
    key: str
    title: str
    description: str
    state: WorkflowStepState
    tone: WorkflowStepTone = "default"
    evidence: str | None = None
    filter_categories: list[HistoryCategory] = Field(default_factory=list)
    workspace_tab: str | None = None


class TaskWorkflowLineageResponse(BaseModel):
    task_id: str
    task_status: str
    annotation_mode: str
    generated_at: datetime
    item_counts: TaskItemCountsRead
    open_escalations: int
    steps: list[WorkflowStepRead]
    current_step_key: str | None = None
    headline: str = ""
    next_action: str = ""
